pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotJoinExchangeNodeInsertRule.java

```diff
@@ -65,14 +62,14 @@ public void onMatch(RelOptRuleCall call) {
 
     RelNode leftExchange;
     RelNode rightExchange;
-    List<RelHint> hints = join.getHints();
-    if (hints.contains(PinotRelationalHints.USE_BROADCAST_DISTRIBUTE)) {
-      // TODO: determine which side should be the broadcast table based on table metadata
-      // TODO: support SINGLETON exchange if the non-broadcast table size is small enough to stay local.
-      leftExchange = LogicalExchange.create(leftInput, RelDistributions.RANDOM_DISTRIBUTED);
+    JoinInfo joinInfo = join.analyzeCondition();
+
+    if (joinInfo.leftKeys.isEmpty()) {
+      // when there's no JOIN key, use broadcast.
+      leftExchange = LogicalExchange.create(leftInput, RelDistributions.SINGLETON);
```

Comment:
just to make sure that I'm understanding this correctly - using `SINGLETON` on the left and `BROADCAST_DISTRIBUTED` on the right would mean that each row in the right table is broadcasted to every node that hosts a segment of the left table, and the left tables remain in place (essentially not exchanged at all)?

> when there's no join key, use broadcast assuming the right table is smaller.

I noticed that there's a `computeSelfCost` on `RelNode` which wires down to basically getting a row count. Is there a way for us to hook that in to make sure that we're actually choosing the smaller table? Obviously this can be left for a follow-up, just wanted to learn :)

Reply:
cost is not enabled. it is either zero or infinite right now. and yes the assumption is correct.